ml_perf/reference_implementation.py

```diff
@@ -103,21 +133,97 @@ def checked_run(name, *cmd):
   # arguments to the actual subprocess because of a quirk in how unknown flags
   # are handled: unknown flags in flagfiles are silently ignored, while unknown
   # flags on the command line will cause the subprocess to abort.
-  logging.info(
-      'Running %s:\n  %s  %s', name, cmd[0], '  '.join(expand_flags(*cmd)))
-
+  cmd, num_instance = expand_flags(*cmd)
+  logging.info('Running %s*%d:\n  %s', name, num_instance, '\n  '.join(cmd))
   with utils.logged_timer('%s finished' % name.capitalize()):
-    completed_process = subprocess.run(
-        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
-    if completed_process.returncode:
-      logging.error('Error running %s: %s', name,
-                    completed_process.stdout.decode())
-      raise RuntimeError('Non-zero return code executing %s' % ' '.join(cmd))
-  return completed_process
-
-
-def get_lines(completed_process, slice):
-  return '\n'.join(completed_process.stdout.decode()[:-1].split('\n')[slice])
+    # if num_instance == 0, use default behavior for GPU
```

Comment:
The responsibility of `checked_run` is to run a subprocess, check that it ran successfully, and return the subprocess' output. It shouldn't changing flags or environment variables. The caller of `checked_run` should be the one containing the `--instance_id` logic, perhaps with a helper function shared by the `selfplay`, `eval, etc functions.

I think you could simplify this code a lot by using Python's `asyncio` library. As an example:

```
import asyncio


# Launch a command as a coroutine and return the stdout output as a string.
async def _exec(*cmd):
    p = await asyncio.create_subprocess_exec(
        *cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.STDOUT)
    chunks = []
    while True:
        chunk = await p.stdout.read(16 * 1024)
        if not chunk:
            break
        chunks.append(chunk.decode())
    return ''.join(chunks)


coroutines = []
for i in range(10):
    coroutines.append(_exec('echo', '#' * 100000))
loop = asyncio.get_event_loop()
results = loop.run_until_complete(asyncio.gather(*coroutines))
loop.close()
for r in results:
    print(r)
```